src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -169,21 +170,22 @@ protected function refreshUser(TokenInterface $token)
 
             try {
                 $refreshedUser = $provider->refreshUser($user);
-                $token->setUser($refreshedUser);
+                $newToken = clone $token;
```

Comment:
`clone` is not always possible. I suppose that we won't have any issues here, but just wanted to point it out.

Reply:
I think we should not clone here. We can just keep the refreshed user and replace the one in the token after all user providers have been processed.

Reply:
I'm not sure to understand what you mean.
Are you suggesting to call `$token->setUser()` only once with the user returned by the last user provider of the loop? The right ("unchanged") user could be in the middle so we need to compare them all, right?
I mean, the "authenticated" state is part of the token. If we don't call `setUser()` on the _original_ token (which contains the original user) then we can't call `isAuthenticated()` and break the loop in case it returns true as done below.

Reply:
You are right here. I forgot that we need to check the token with the refreshed user.

But since the `TokenInterface` is serializable we could maybe stick to `$newToken = unserialize(serialize($token));` instead. This should always be safe, right?

Reply:
Probably more expensive in term of memory and CPU time but feels a bit more robust. 
@nicolas-grekas any thought?

Reply:
You mean to don't call setUser here ? In that case, the token would stay authenticated even if the refreshed user has changed. So we have to manually compare the user stored in the token with the refreshed user here ?

Reply:
An alternative to clone could be to turn the private `AbstractToken::hasUserChanged()` method into an internal public static one which could then be used by both AbstractToken to perform checks against its own internal user as it does currently, and here to compare each loaded user with the original one.

Reply:
@nicolas-grekas Can you give your input here?

Reply:
@chalasr I like this idea but why static ? As you would need to compare the user stored in the token, you would need to use `$this->user` no ?

Reply:
@gpekz I forgot to mention the signature change: 
`public static function hasUserChanged(UserInterface $originalUser, UserInterface $user);` so that `AbstractToken` would pass `$this->user` as `$originalUser` for its internal use and we could use it externally without creating dummy token instances

Reply:
actually we need to move this part https://github.com/symfony/security-core/blob/master/Authentication/Token/AbstractToken.php#L87-L99 in `hasUserChanged()` and remove any UserInterface typehint from the signature in order to handle strings (and objects implementing `__toString()`, see https://github.com/symfony/security-core/blob/master/Authentication/Token/TokenInterface.php#L60)

Reply:
But tokens that do not extend AbstractToken use their own `setUser` and `isAuthenticated` methods here. Doing this change will break it. Isn't it a problem ?

Reply:
@gpekz I think you're right, it's not good to rely on hasUserChanged as it's AbstractToken specific. 
Compared clone to serialize on 10 providers with 9/10 failing to refresh and got no significative difference https://blackfire.io/profiles/compare/5f7a76a9-7786-4bd0-b360-96e43ad07b1b/graph.

+1 for using `unserialize(serialize($newToken))` instead of `clone $newToken` given the token is already safely unserialized beforehand.

Reply:
Done :)